ast/compile.go

```diff
@@ -2212,7 +2212,7 @@ func resolveRef(globals map[Var]Ref, ignore *declaredVarStack, ref Ref) Ref {
 			} else {
 				r = append(r, x)
 			}
-		case Ref, Array, Object, Set:
+		case Ref, Array, Object, Set, *ArrayComprehension, *SetComprehension, *ObjectComprehension, Call:
```

Comment:
It would be good to include a unit test in the ast package for this case. I'd probably add one for Calls and one for the comprehensions. I notice this fixes an issue whereby comprehensions would not have been resolved if they were embedded in refs, e.g., `p[[x | q = x]]` where `q` was global.